man/vkill.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VKILL "8" "May 2013" "vkill  -- sends signals to processes within other contexts" "System Administration"
.SH NAME
vkill \- sends signals to processes within other contexts
.SH SYNOPSIS
.B vkill
[\fI--xid|-c <xid>\fR] [\fI-s <signal>\fR] [\fI--\fR] \fI<pid>*\fR
.SH DESCRIPTION
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2003 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vkill.c
// $Id$    --*- c -*--

// Copyright (C) 2003 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "vserver.h"
#include "util.h"

#include "lib/virtual.h"

#include <getopt.h>
#include <signal.h>
#include <stdlib.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <sys/wait.h>

#define ENSC_WRAPPERS_VSERVER	1
#define ENSC_WRAPPERS_UNISTD	1
#include <wrappers.h>

#define CMD_HELP	0x8000
#define CMD_VERSION	0x8001

int		wrapper_exit_code = 1;

static struct option const
CMDLINE_OPTIONS[] = {
  { "help",     no_argument,        0, CMD_HELP },
  { "version",  no_argument,        0, CMD_VERSION },
  { "xid",      required_argument,  0, 'c' },
  { 0,0,0,0 }
};

struct Arguments
{
    xid_t		xid;
    int			sig;
};

static char const * const SIGNALS[] = {
  // 0      1      2          3       4       5       6       7
  "UNUSED", "HUP", "INT",     "QUIT", "ILL",  "TRAP", "ABRT", "UNUSED",
  "FPE",    "KILL", "USR1",   "SEGV", "USR2", "PIPE", "ALRM", "TERM",
  "STKFLT", "CHLD", "CONT",   "STOP", "TSTP", "TTIN", "TTOU", "IO",
  "XCPU",   "XFSZ", "VTALRM", "PROF", "WINCH",
  0,
};

static void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " [--xid|-c <xid>] [-s <signal>] [--] <pid>*\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");
  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vkill " VERSION " -- sends signals to processes within other contexts\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2003 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static int
str2sig(char const *str)
{
  char	*errptr;
  int	res = strtol(str, &errptr, 10);
  
  if (*errptr!='\0') res=-1;
  if (res==-1 && strncmp(str,"SIG",3)==0) str+=3;
  if (res==-1) {
    char const * const	*ptr = SIGNALS;
    for (;*ptr!=0; ++ptr) {
      if (strcmp(*ptr,str)!=0) continue;
      res = ptr-SIGNALS;
      break;
    }
  }

  return res;
}

#if defined(VC_ENABLE_API_LEGACY)
inline static ALWAYSINLINE int
kill_wrapper_legacy(xid_t UNUSED xid, char const *proc, int UNUSED sig)
{
  pid_t		pid;

  signal(SIGCHLD, SIG_DFL);
  pid = Efork();

  if (pid>0) {
    int		status;
    int		res;
    while ((res=wait4(pid, &status, 0,0))==-1 &&
	   (errno==EAGAIN || errno==EINTR)) {}

    return (res==0 && WIFEXITED(status) && WEXITSTATUS(status)) ? 0 : 1;
  }

  execl(LEGACYDIR "/vkill", "legacy/vkill", proc, (void *)(0));
  perror("vkill: execl()");
  exit(1);
}
#endif

inline static int
kill_wrapper(xid_t xid, char const *pid_s, int sig)
{
  pid_t	pid;
  long	tmp;

  if (!isNumber(pid_s, &tmp, true)) {
    WRITE_MSG(2, "vkill: '");
    WRITE_STR(2, pid_s);
    WRITE_MSG(2, "' is not a number\n");
  }
  pid = (pid_t) tmp;

  if (xid==VC_NOCTX)
    xid = vc_get_task_xid(pid);
  if (vc_ctx_kill(xid,pid,sig)==-1) {
#if defined(VC_ENABLE_API_LEGACY)
    int		err = errno;
    if (vc_get_version(VC_CAT_COMPAT)==-1)
      return kill_wrapper_legacy(xid, pid_s, sig);
    else {
      errno = err;
#else
    {
#endif
      perror("vkill: vc_ctx_kill()");
      return 1;
    }
  }
  return 0;
}


int main(int argc, char *argv[])
{
  int			fail = 0;
  struct Arguments	args = {
    .xid = VC_NOCTX,
    .sig = SIGTERM,
  };
  
  while (1) {
    int		c = getopt_long(argc, argv, "c:s:", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0], 0);
      case CMD_VERSION	:  showVersion();
      case 'c'		:  args.xid = Evc_xidopt2xid(optarg,true); break;
      case 's'		:  args.sig = str2sig(optarg);             break;
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return EXIT_FAILURE;
	break;
    }
  }

  if (args.sig==-1) {
    WRITE_MSG(2, "Invalid signal specified\n");
    return EXIT_FAILURE;
  }

  if (args.xid==VC_NOCTX && optind==argc) {
    WRITE_MSG(2, "No pid specified\n");
    return EXIT_FAILURE;
  }

  if (optind==argc)
    fail += kill_wrapper(args.xid, "0", args.sig);
  else for (;optind<argc;++optind)
    fail += kill_wrapper(args.xid, argv[optind], args.sig);

  return fail==0 ? EXIT_SUCCESS : EXIT_FAILURE;
}

#ifdef TESTSUITE
void
vkill_test()
{
  assert(str2sig("0") ==0 );
  assert(str2sig("1") ==1 );
  assert(str2sig("10")==10);
  assert(str2sig("SIGHUP")==1);
  assert(str2sig("HUP")   ==1);
  assert(str2sig("SIGCHLD")==17);
  assert(str2sig("CHLD")   ==17);
  assert(str2sig("x")==-1);
  assert(str2sig("1 0")==-1);

  return 0;
}
#endif
